src/hash.jl

```diff
@@ -0,0 +1,80 @@
+# The Base implementation of hash(::Real) is based on Base.decompose
+# to guarantee that all numbers that compare equal are given the same
+# hash. We implement Base.decompose for Mag and Arf.
+function Base.decompose(x::Union{mag_struct,Ptr{mag_struct}})::Tuple{UInt,BigInt,Int}
+    isinf(x) && return 1, 0, 0
+
+    if x isa Ptr{mag_struct}
+        x = unsafe_load(x)
+    end
+
+    pow = BigInt()
+    ccall(@libflint(fmpz_get_mpz), Nothing, (Ref{BigInt}, Ref{UInt}), pow, x.exponent)
```

Comment:
this is just a side-comment again: maybe we should consider replacing all the `ccall`s + our lib-macros by the new `@ccall` syntax? (if so I'll do it in a separate PR)
e.g. this would become
```julia
@ccall libflint.fmpz_get_mpz(pow::Ref{BigInt}, x.exponent::Ref{UInt})::Cvoid
```
if we replace `@libflint` by simple `const libflint = Arb_jll.libflint`